src/librbd/ImageCtx.cc

```diff
@@ -803,4 +809,98 @@ namespace librbd {
       async_requests_cond.Wait(async_ops_lock);
     }
   }
+
+  bool ImageCtx::_filter_metadata_confs(const string &prefix, map<string, bufferlist> &pairs,
+                                        map<string, bufferlist> *res) {
+    size_t conf_prefix_len = prefix.size();
+
+    const std::vector<string> configs = boost::assign::list_of(
+        "rbd_cache")(
+        "rbd_cache_writethrough_until_flush")(
+        "rbd_cache_size")(
+        "rbd_cache_max_dirty")(
+        "rbd_cache_target_dirty")(
+        "rbd_cache_max_dirty_age")(
+        "rbd_cache_max_dirty_object")(
+        "rbd_cache_block_writes_upfront")(
+        "rbd_concurrent_management_ops")(
+        "rbd_balance_snap_reads")(
+        "rbd_localize_snap_reads")(
+        "rbd_balance_parent_reads")(
+        "rbd_localize_parent_reads")(
+        "rbd_readahead_trigger_requests")(
+        "rbd_readahead_max_bytes")(
+        "rbd_readahead_disable_after_bytes")(
+        "rbd_clone_copy_on_read")(
+        "rbd_blacklist_on_break_lock")(
+        "rbd_blacklist_expire_seconds")(
+        "rbd_request_timed_out_seconds");
+
+    string start = prefix;
+    for (map<string, bufferlist>::iterator it = pairs.begin(); it != pairs.end(); ++it) {
+      if (it->first.size() <= conf_prefix_len || it->first.compare(0, conf_prefix_len, prefix))
+        return false;
+
+      string key = it->first.substr(conf_prefix_len, it->first.size() - conf_prefix_len);
+      for (std::vector<string>::const_iterator cit = configs.begin();
+           cit != configs.end(); ++cit) {
+        if (!key.compare(*cit)) {
+          res->insert(make_pair(key, it->second));
+          break;
+        }
+      }
+    }
+    return true;
+  }
+
+  void ImageCtx::apply_metadata_confs() {
+    ldout(cct, 20) << __func__ << dendl;
+    static uint64_t max_conf_items = 128;
+
+    string start = METADATA_CONF_PREFIX;
+    int r = 0, j = 0;
+    bool is_continue;
+    do {
+      map<string, bufferlist> pairs, res;
+      r = cls_client::metadata_list(&md_ctx, header_oid, start, max_conf_items, &pairs);
+      if (r < 0) {
+        lderr(cct) << __func__ << " couldn't list conf metadatas: " << r << dendl;
+        break;
+      }
+      if (pairs.empty())
+        break;
+      
+      is_continue = _filter_metadata_confs(METADATA_CONF_PREFIX, pairs, &res);
+      for (map<string, bufferlist>::iterator it = res.begin(); it != res.end(); ++it) {
+        j = cct->_conf->set_val(it->first.c_str(), it->second.c_str());
```

Comment:
it's not just the apply that will affect other images with the same cct - set_val will too